overlord/snapstate/autorefresh_gating_test.go

```diff
@@ -211,6 +223,390 @@ slots:
+			"48h", // max duration
+			"24h", // expected
+		},
+		{
+			"2021-05-03T10:00:00Z", // last refreshed (1 month ago)
+			"2021-06-01T10:00:00Z", // first held (2 days ago)
+			"48h", // max duration
+			"00h", // expected
+		},
+		{
+			"2021-03-08T10:00:00Z", // last refreshed (almost 3 months ago)
+			"2021-06-01T10:00:00Z", // first held
+			"2160h",                // max duration (90 days)
+			"72h",                  // expected
+		},
+		{
+			"2021-03-04T10:00:00Z", // last refreshed
+			"2021-06-01T10:00:00Z", // first held (2 days ago)
+			"2160h",                // max duration (90 days)
+			"-24h",                 // expected (refresh is 1 day overdue)
+		},
+		{
+			"2021-06-01T10:00:00Z", // last refreshed (2 days ago)
+			"2021-06-03T10:00:00Z", // first held now
+			"2160h",                // max duration (90 days)
+			"2112h",                // expected (max minus 2 days)
+		},
+	} {
+		lastRefresh, err := time.Parse(time.RFC3339, tc.lastRefresh)
+		c.Assert(err, IsNil)
+		firstHeld, err := time.Parse(time.RFC3339, tc.firstHeld)
+		c.Assert(err, IsNil)
+		maxDuration, err := time.ParseDuration(tc.maxDuration)
+		c.Assert(err, IsNil)
+		expected, err := time.ParseDuration(tc.expected)
+		c.Assert(err, IsNil)
+
+		left := snapstate.HoldDurationLeft(now, lastRefresh, firstHeld, maxDuration, maxPostponement)
+		c.Check(left, Equals, expected, Commentf("case #%d", i))
+	}
+}
+
+func (s *autorefreshGatingSuite) TestLastRefreshedHelper(c *C) {
+	st := s.state
+	st.Lock()
+	defer st.Unlock()
+
+	inf := mockInstalledSnap(c, st, snapAyaml, false)
+	stat, err := os.Stat(inf.MountFile())
+	c.Assert(err, IsNil)
+
+	refreshed, err := snapstate.LastRefreshed(st, "snap-a")
+	c.Assert(err, IsNil)
+	c.Check(refreshed, DeepEquals, stat.ModTime())
+
+	t, err := time.Parse(time.RFC3339, "2021-01-01T10:00:00Z")
+	c.Assert(err, IsNil)
+
+	var snapst snapstate.SnapState
+	c.Assert(snapstate.Get(st, "snap-a", &snapst), IsNil)
+	snapst.LastRefreshTime = &t
+	snapstate.Set(st, "snap-a", &snapst)
+
+	refreshed, err = snapstate.LastRefreshed(st, "snap-a")
+	c.Assert(err, IsNil)
+	c.Check(refreshed, DeepEquals, t)
+}
+
+func (s *autorefreshGatingSuite) TestHoldRefreshHelper(c *C) {
+	st := s.state
+	st.Lock()
+	defer st.Unlock()
+
+	restore := snapstate.MockTimeNow(func() time.Time {
+		t, err := time.Parse(time.RFC3339, "2021-05-10T10:00:00Z")
+		c.Assert(err, IsNil)
+		return t
+	})
+	defer restore()
+
+	mockInstalledSnap(c, st, snapAyaml, false)
+	mockInstalledSnap(c, st, snapByaml, false)
+	mockInstalledSnap(c, st, snapCyaml, false)
+	mockInstalledSnap(c, st, snapDyaml, false)
+	mockInstalledSnap(c, st, snapEyaml, false)
+	mockInstalledSnap(c, st, snapFyaml, false)
+
+	mockLastRefreshed(c, st, "2021-05-09T10:00:00Z", "snap-a", "snap-b", "snap-c", "snap-d", "snap-e", "snap-f")
+
+	c.Assert(snapstate.HoldRefresh(st, "snap-a", 0, "snap-b", "snap-c"), IsNil)
+	c.Assert(snapstate.HoldRefresh(st, "snap-a", 0, "snap-e"), IsNil)
```

Comment:
shouldn't this call be merged with the previous? if there's a reason to keep it separate maybe there should be a comment

Reply:
That's not required, this is to demonstrate that multiple calls with same gating snap is fine. I'll add a comment.